nupic/research/frameworks/pytorch/distributed_sampler.py

```diff
@@ -0,0 +1,108 @@
+    allowing the number of batches per process to be off-by-one from each other.
+    This makes this sampler usable for validation (it's deterministic and
+    doesn't require shuffling). Diffs to the code are annotated with "MODIFY"
+    comments.
+
+    Sampler that restricts data loading to a subset of the dataset.
+    It is especially useful in conjunction with
+    :class:`torch.nn.parallel.DistributedDataParallel`. In such case, each
+    process can pass a DistributedSampler instance as a DataLoader sampler,
+    and load a subset of the original dataset that is exclusive to it.
+    .. note::
+        Dataset is assumed to be of constant size.
+    Arguments:
+        dataset: Dataset used for sampling.
+        num_replicas (optional): Number of processes participating in
+            distributed training.
+        rank (optional): Rank of the current process within num_replicas.
+        shuffle (optional): If true (default), sampler will shuffle the indices
+    .. warning::
+        In distributed mode, calling the ``set_epoch`` method is needed to
+        make shuffling work; each process will use the same random seed
+        otherwise.
+    Example::
+        >>> sampler = DistributedSampler(dataset) if is_distributed else None
+        >>> loader = DataLoader(dataset, shuffle=(sampler is None),
+        ...                     sampler=sampler)
+        >>> for epoch in range(start_epoch, n_epochs):
+        ...     if is_distributed:
+    """
+
+    def __init__(self, dataset, num_replicas=None, rank=None, shuffle=True):
+        if num_replicas is None:
+            if not dist.is_available():
+                raise RuntimeError("Requires distributed package to be available")
+            num_replicas = dist.get_world_size()
+        if rank is None:
+            if not dist.is_available():
+                raise RuntimeError("Requires distributed package to be available")
+            rank = dist.get_rank()
+        self.dataset = dataset
+        self.num_replicas = num_replicas
+        self.rank = rank
+        self.epoch = 0
+
+        # BEGIN MODIFY: allow different num_samples values in different processes.
+        # self.num_samples = int(math.ceil(len(self.dataset) * 1.0 / self.num_replicas))
+        # self.total_size = self.num_samples * self.num_replicas
+        self.num_samples = len(range(self.rank, len(self.dataset), self.num_replicas))
+        self.total_size = len(self.dataset)
+        # END MODIFY
+
+        self.shuffle = shuffle
+
+    def __iter__(self):
+        if self.shuffle:
+            # deterministically shuffle based on epoch
+            g = torch.Generator()
+            g.manual_seed(self.epoch)
+            indices = torch.randperm(len(self.dataset), generator=g).tolist()
+        else:
+            indices = list(range(len(self.dataset)))
+
+        # BEGIN MODIFY: don't pad the data.
+        # # add extra samples to make it evenly divisible
```

Comment:
Do you know why was that a requirement for the original sampler?

Reply:
Good question, I'm adding some text to the docstring and an assert to explain how this class can be error-prone. I imagine the original sampler had this requirement to avoid these potential issues.